### PR TITLE
Refactor publishing path, add diagnostics, update documentation (#1)

### PR DESCRIPTION
**Pull request by hansgschossmann**
* Refactor publishing path, add diagnostics, update documentation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -330,6 +351,177 @@ In certain use cases it may make sense to read configuration information from or
...
+        current working set in MB: 89
+        ======================================================================
+
+This configuration batches as much OPC node value udpates as possible. The maximum IoTHub message size is 256 kB, which is configured here. There is no send interval requested, which makes the time when data is ingested
+completly controlled by the data itself. This configuration has the least probability of loosing any OPC node values and could be used for publishing a high number of nodes.
+When using this configuration you need to ensure, that your scenario does not have conditions where high latency is introduced (because the message size of 256 kB is not reached).
+
 # Debugging the Application
 
 ## Native on Winodws
```
Comment (khilscher) [inline README.md]: Typo...Winodws #Resolved

File: README.md [hunk 2]
```diff
@@ -330,6 +351,177 @@ In certain use cases it may make sense to read configuration information from or
...
+        time in ms for failed msgs: 0
+        avg time in ms for failed msg: 0
+        messages too large to sent to IoTHub: 0
+        times we missed send interval: 0
+        ---------------------------------
+        current working set in MB: 89
+        ======================================================================
+
+This configuration batches as much OPC node value udpates as possible. The maximum IoTHub message size is 256 kB, which is configured here. There is no send interval requested, which makes the time when data is ingested
+completly controlled by the data itself. This configuration has the least probability of loosing any OPC node values and could be used for publishing a high number of nodes.
```
Comment (khilscher) [inline README.md]: Typo...completly  #Resolved

**Review by khilscher — APPROVED** [review]

File: src/Diagnostics.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            while (true)
+            {
+                try
+                {
+
+                    if (cancellationtoken.IsCancellationRequested)
+                    {
+                        return;
+                    }
+                    await Task.Delay((int)_diagnosticsInterval * 1000);
```
Comment (mregen) [inline src/Diagnostics.cs]: I'd pass the cancellation token to the Delay function and check it after delay.. #Resolved

File: src/Diagnostics.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            get => _diagnosticsInterval;
+            set => _diagnosticsInterval = value;
+        }
+        private static uint _diagnosticsInterval = 0;
+
+        public static int IotHubMessagingMessagesSentCount
+        {
+            get => _messagesSentCount;
+        }
+        private static int _messagesSentCount = 0;
```
Comment (mregen) [inline src/Diagnostics.cs]: would be cleaner to put all private static in a block at the bottom... (I know you don't like regions, but...) #Resolved

File: src/Diagnostics.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    {
+
+        public static uint DiagnosticsInterval
+        {
+            get => _diagnosticsInterval;
+            set => _diagnosticsInterval = value;
+        }
+        private static uint _diagnosticsInterval = 0;
+
+        public static int IotHubMessagingMessagesSentCount
```
Comment (mregen) [inline src/Diagnostics.cs]: should be possible to write this as:
public static int IotHubMessagingMessagesSentCount => _messagesSentCount; ?
 #Resolved

File: src/Diagnostics.cs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+
+        }
+        public async static Task Shutdown()
+        {
+            // wait for diagnostic task completion if it is enabled
+            if (_showDiagnosticsInfoTask != null)
+            {
+                _shutdownTokenSource.Cancel();
+                await _showDiagnosticsInfoTask;
```
Comment (mregen) [inline src/Diagnostics.cs:48]: null or dispose the static memebers or they are never freed #Resolved

File: src/IotHubMessaging.cs [hunk 7]
```diff
@@ -36,12 +36,13 @@ public static Microsoft.Azure.Devices.Client.TransportType IotHubProtocol
...
+        public const uint IotHubMessageSizeMax = (256 * 1024);
+        public static uint IotHubMessageSize
         {
-            get => _maxSizeOfIoTHubMessageBytes;
-            set => _maxSizeOfIoTHubMessageBytes = value;
+            get => _iotHubMessageSize;
+            set => _iotHubMessageSize = value;
         }
-        private static uint _maxSizeOfIoTHubMessageBytes = 4096;
+        private static uint _iotHubMessageSize = 4096;
```
Comment (mregen) [inline src/IotHubMessaging.cs]: 4096 would be the 'meter size'? #Resolved

File: src/IotHubMessaging.cs [hunk 8]
```diff
@@ -66,6 +67,88 @@ public static string IotDeviceCertStorePath
...
+        private static long _enqueueFailureCount;
+
+        public static long DequeueCount
+        {
+            get => _dequeueCount;
+        }
+        private static long _dequeueCount;
+
+
+        public static long MissedSendIntervalCount
```
Comment (mregen) [inline src/IotHubMessaging.cs]: beauty contest: all public static could be one liners and all privates in a block... #Resolved

File: src/IotHubMessaging.cs [hunk 9]
```diff
@@ -83,26 +166,16 @@ private class OpcUaValue
...
         /// Ctor for the class.
         /// </summary>
         public IotHubMessaging()
         {
-            _sendQueue = new ConcurrentQueue<string>();
-            _sendQueueEvent = new AutoResetEvent(false);
-            _messageList = new List<OpcUaMessage>();
-            _messageListSemaphore = new SemaphoreSlim(1);
-            _currentSizeOfIotHubMessageBytes = 0;
+            _monitoredItemsDataQueue = new BlockingCollection<string>(_monitoredItemsQueueCapacity);
```
Comment (mregen) [inline src/IotHubMessaging.cs]: This is weird: the static data queue is initialized in the constructor? Shouldn't it be done once and for all? #Resolved

File: src/IotHubMessaging.cs [hunk 10]
```diff
@@ -218,172 +291,204 @@ public IotHubMessaging()
             try
             {
                 _tokenSource.Cancel();
-                await _dequeueAndSendTask;
+                await _monitoredItemsProcessorTask;
```
Comment (mregen) [inline src/IotHubMessaging.cs:229]: null static vars if possible #Resolved

File: src/IotHubMessaging.cs [hunk 11]
```diff
@@ -218,172 +291,204 @@ public IotHubMessaging()
...
+        /// <summary>
+        /// Enqueue a message for sending to IoTHub.
+        /// </summary>
+        /// <param name="json"></param>
         public void Enqueue(string json)
         {
-            _sendQueue.Enqueue(json);
-            _sendQueueEvent.Set();
+            // Try to add the message.
+            _enqueueCount++;
```
Comment (mregen) [inline src/IotHubMessaging.cs]: Shouldn't this be interlocked? Multiple threads calling at the same time? #Resolved

**Review by mregen — DISMISSED** [review]
Not sure about the constructor and concurrency of counters. Otherwise just cosmetics.

File: README.md (hunk 1 above)
Comment (hansgschossmann) [reply README.md]: changed

---
In reply to: [146973971](https://github.com/Azure/iot-edge-opc-publisher/pull/85#discussion_r146973971) [](ancestors = 146973971)

File: src/Diagnostics.cs (hunk 3 above)
Comment (hansgschossmann) [reply src/Diagnostics.cs]: good idea

---
In reply to: [147062995](https://github.com/Azure/iot-edge-opc-publisher/pull/85#discussion_r147062995) [](ancestors = 147062995)

File: src/Diagnostics.cs (hunk 4 above)
Comment (hansgschossmann) [reply src/Diagnostics.cs]: yeah, i saw that once in the stack and i like it too. i will make the change in the whole solution.


---
In reply to: [147063206](https://github.com/Azure/iot-edge-opc-publisher/pull/85#discussion_r147063206) [](ancestors = 147063206)

File: src/IotHubMessaging.cs (hunk 7 above)
Comment (hansgschossmann) [reply src/IotHubMessaging.cs]: yes

---
In reply to: [147064343](https://github.com/Azure/iot-edge-opc-publisher/pull/85#discussion_r147064343) [](ancestors = 147064343)

File: src/Diagnostics.cs (hunk 5 above)
Comment (hansgschossmann) [reply src/Diagnostics.cs]: i have beatified it.

---
In reply to: [147063571](https://github.com/Azure/iot-edge-opc-publisher/pull/85#discussion_r147063571) [](ancestors = 147063571)

File: src/IotHubMessaging.cs (hunk 10 above)
Comment (hansgschossmann) [reply src/IotHubMessaging.cs:229]: done

---
In reply to: [147065955](https://github.com/Azure/iot-edge-opc-publisher/pull/85#discussion_r147065955) [](ancestors = 147065955)

File: src/IotHubMessaging.cs (hunk 8 above)
Comment (hansgschossmann) [reply src/IotHubMessaging.cs]: check the next iteration, if i have chance to win 

---
In reply to: [147065037](https://github.com/Azure/iot-edge-opc-publisher/pull/85#discussion_r147065037) [](ancestors = 147065037)

File: src/IotHubMessaging.cs (hunk 9 above)
Comment (hansgschossmann) [reply src/IotHubMessaging.cs]: moved to init.

---
In reply to: [147065741](https://github.com/Azure/iot-edge-opc-publisher/pull/85#discussion_r147065741) [](ancestors = 147065741)

File: src/IotHubMessaging.cs (hunk 11 above)
Comment (hansgschossmann) [reply src/IotHubMessaging.cs]: thx

---
In reply to: [147066124](https://github.com/Azure/iot-edge-opc-publisher/pull/85#discussion_r147066124) [](ancestors = 147066124)

File: README.md (hunk 2 above)
Comment (hansgschossmann) [reply README.md]: changed

---
In reply to: [146974139](https://github.com/Azure/iot-edge-opc-publisher/pull/85#discussion_r146974139) [](ancestors = 146974139)

File: src/Diagnostics.cs (hunk 6 above)
Comment (hansgschossmann) [reply src/Diagnostics.cs:48]: done

---
In reply to: [147063991](https://github.com/Azure/iot-edge-opc-publisher/pull/85#discussion_r147063991) [](ancestors = 147063991)

File: src/OpcSession.cs [hunk 12]
```diff
@@ -889,7 +937,7 @@ private void InternalDisconnectAsync()
                 }
 
                 // Stop publishing.
```
Comment (mregen) [inline src/OpcSession.cs]: >// Stop publishing. [](start = 16, length = 19)

typo?? #Resolved

**Review by mregen — APPROVED** [review]
:shipit:

File: src/OpcSession.cs (hunk 12 above)
Comment (hansgschossmann) [reply src/OpcSession.cs]: changed

---
In reply to: [147425944](https://github.com/Azure/iot-edge-opc-publisher/pull/85#discussion_r147425944) [](ancestors = 147425944)

**Review by barnstee — COMMENTED** [review]
Looks good.

**Pull request merged**